# Application to causal structure discovery

See [the paper about the technique](https://arxiv.org/pdf/1609.00672.pdf). Read in particular Section I, section II can be skimmed, as well as III.A. The main idea can be understood by reading III.B; we'll use Example 1 in our description below; read the pen-and-paper proof in Example 1, we'll formulate it as a linear program.

Optional, read III.C, skimming through the technical definitions, pay attention to Example 4. The use of linear programming is detailed in Section IV, particularly IV.B. Skip the rest. Appendix A provides the data we need.

## An example of causal structure

Consider the following causal structure:

![scen15DAG.svg](scen15DAG.svg)

where we observe the variables $A$, $B$ and $C$ that take the values $a,b,c=0,1$. We observe the following correlations:

\begin{equation}
P_{ABC}(a,b,c) = \begin{cases} 1/2, & \text{if } a=b=c, \\ 0, & \text{otherwise}. \end{cases}
\end{equation}

We try to understand whether this perfectly correlated distribution can arise in a causal structure where the variables $A$, $B$ and $C$ only depend on information that is shared with another party only.



### A causal model

Thus, there are unobserved variables $X$, $Y$, $Z$, with distributions $P_X(x)$, $P_Y(y)$, $P_Z(z)$, such that the variable $A$ is fully described by $P_{A|XY}(a|x,y)$, the variable $B$ by $P_{B|YZ}(b|y,z)$ and the variable $C$ by $P_{C|XZ}(c|x,z)$, and we have

\begin{equation}
\label{Eq:Model}
P_{ABC} (a,b,c) = \sum_{xyz} P_X(x) P_Y(y) P_Z(z) P_{A|XY}(a|x,y) P_{B|YZ}(b|y,z) P_{C|XZ}(c|x,z).
\end{equation}

Now, assume that the variables $x,y,z$ are integers between $0$ and $N$. Then, testing if our $P_{ABC}$ has a model of the form \eqref{Eq:Model} would be a polynomial feasibility problem (and a hard one!, already for $N\ge3$). But we do not even know the type of the unobserved variables $X$, $Y$, $Z$ (still, see https://arxiv.org/abs/1709.00707 ).

### Test using the "inflation technique" which maps to LP

We will use another method, amenable to linear programming. We will make a (numerical) proof by contradication. Assume that $P_{ABC}$ has a model of the form \eqref{Eq:Model}. Then, we can imagine a variation on that model, where we duplicate the variable $Y$, and wire the relations between the variables a bit differently.

![scen15InflationDAGV3.svg](scen15InflationDAGV3.svg)
![TriDagSubA2B1C1.svg](TriDagSubA2B1C1.svg)

This is called an *inflated* scenario. There, we obtain the slightly different correlations:

\begin{equation}
\label{Eq:Model1}
P_{A_2B_1C_1} (a_2,b_1,c_1) = \sum_{x_1 y_1 y_2 z} P_{X_1}(x_1) P_{Y_1}(y_1) P_{Y_2}(y_2) P_{Z_1}(z_1) P_{A_2|X_1 Y_2}(a_2|x_1,y_2) P_{B_1|Y_1 Z_1}(b_1|y_1,z_1) P_{C_1|X_1 Z_1}(c_1|x_1,z_1).
\end{equation}

Note that, however, the marginal distribution of the inflated correlations

\begin{equation}
P_{A_2 C_1} = \sum_{b_1} P_{A_2B_1C_1} (a_2,b_1,c_1) = \sum_{x_1 y_2 z} P_{X_1}(x_1) P_{Y_2}(y_2) P_{Z_1}(z_1) P_{A_2|X_1 Y_2}(a_2|x_1,y_2) P_{C_1|X_1 Z_1}(c_1|x_1,z_1)
\end{equation}

has the same form as the marginal distribution of the original scenario

\begin{equation}
P_{AC} (a,c) = \sum_b P_{ABC}(a,b,c) = \sum_{xyz} P_X(x) P_Y(y) P_Z(z) P_{A|XY}(a|x,y) P_{C|XZ}(c|x,z).
\end{equation}


We thus have

\begin{equation}
P_{AC} (i,k) = P_{A_2 C_1}(i,k), \qquad \forall i,k\;.
\end{equation}

The same argument holds for

\begin{equation}
P_{BC} (j,k) = P_{B_1 C_1}(j,k), \qquad \forall j,k\;.
\end{equation}

Now, let us examine $P_{A_2 B_1}(a_2, b_1) = \sum_{c_1} P_{A_2B_1C_1} (a_2,b_1,c_1)$. It corresponds, after removal of $C_1$, to the graph:

![Marginal.svg](Marginal.svg)

where the variables $A_2$ and $B_1$ are independent. We thus have:

\begin{equation}
P_{A_2 B_1}(a_2, b_1) = P_{A_2}(a_2) P_{B_1}(b_1)
\end{equation}

which we can now match with the original problem:

\begin{equation}
P_{A_2 B_1} (i,j) = P_{A}(i) P_{B}(j), \qquad \forall i,j\;.
\end{equation}

## The linear program

Writing all these constraints together, we have ($\forall i,j,k$ is implicit):

\begin{align}
\sum_j P_{A_2 B_1 C_1}(i,j,k) & = \sum_j P_{ABC} (i,j,k), \\
\sum_i P_{A_2 B_1 C_1}(i,j,k) &= \sum_i P_{ABC} (i,j,k), \\
\sum_k P_{A_2 B_1 C_1}(i,j,k) &= P_{A}(i) P_{B}(j), \\
P_{A_2 B_1 C_1}(i,j,k) &\ge 0
\end{align}

Now, the inflated correlations $P_{A_2 B_1 C_1}$ may obey additional constraints, but we remark that the constraints listed above correspond to a linear program in the primal form: indeed, the right-hand side of the equations are constant values that depend only on the coefficients $P_{ABC}(i,j,k)$ which are known.

\begin{equation}
  \begin{array}{rl}
    \text{minimize} & 0 \\
    \text{over} & \vec{v} \in \mathbb{R}^n \\
    & M \vec{v} = \vec{b} \\
    & \vec{v} \ge 0
  \end{array}
\end{equation}

where the objective is trivial, the constraint right-hand side $\vec{b}$ is the only part of the problem that depends on $P_{ABC}$, and the matrix $M$ only depends on the problem structure (matching the marginals).

Now, if this linear program is infeasible, it proves by contradiction that no model exists for the original problem (because original problem has model => inflation has a model).

#### Homework 1

Write the linear program using Convex.jl, and verify if the distribution $P_{ABC}$ is compatible with the inflation (hint: it should not).

You can use the numerically better behaved variant that has the slack variable $z$:

\begin{equation}
\label{Eq:Homework1}
  \begin{array}{rl}
    \text{maximize} & z \\
    \text{over} & z \in \mathbb{R}, \vec{v} \in \mathbb{R}^n \\
    & M \vec{v} = \vec{b} \\
    & \vec{v} \ge z
  \end{array}
\end{equation}

#### Homework 2

For which values of $t$ the following distribution is compatible with the inflated model?

\begin{equation}
P_{ABC}(a,b,c) = \begin{cases} 1/2 t, & \text{if } a=b=c, \\ (1-t)/6, & \text{otherwise}. \end{cases}
\end{equation}

#### Homework 3

Test the distribution:

\begin{equation}
\label{Eq:W}
P_{ABC}(a,b,c) = \begin{cases} 1/3, & \text{if } a+b+c = 1, \\ 0, & \text{otherwise}. \end{cases}
\end{equation}

This distribution should be compatible with the inflation above; nevertheless it is not compatible with the causal structure we test (see Example 2 of the [paper](https://arxiv.org/pdf/1609.00672.pdf)). Why is the linear program feasible then?

#### Homework 4

Solve one of the following questions:

- Consider the dual problem of \eqref{Eq:Homework1}. How to interpret the dual variables and the dual objective? Read the part about infeasibility certificates in the [Mosek Cookbook, section 2.3](https://docs.mosek.com/modeling-cookbook/linear.html). Can you derive a causal compatibility inequality as in Example 4 of the [paper](https://arxiv.org/pdf/1609.00672.pdf))?

- Implement the Spiral inflation given in FIG. 3 of the [paper](https://arxiv.org/pdf/1609.00672.pdf), and verify that the distribution \eqref{Eq:W} is incompatible.


## Homework 1

In [28]:
# HW1
using Convex
using GLPK
using Printf

# variable to optimize
v = Variable(8)
# slack variable
z = Variable(1)
# Contraints (taken from the paper)
b = [1/2, 0, 0, 1/2, 1/2, 0, 0, 1/2, 1/4, 1/4, 1/4, 1/4]
M = [1 1 0 0 0 0 0 0;
     0 0 1 1 0 0 0 0;
     0 0 0 0 1 1 0 0;
     0 0 0 0 0 0 1 1;
     1 0 1 0 0 0 0 0;
     0 1 0 1 0 0 0 0;
     0 0 0 0 1 0 1 0;
     0 0 0 0 0 1 0 1;
     1 0 0 0 1 0 0 0;
     0 1 0 0 0 1 0 0;
     0 0 1 0 0 0 1 0;
     0 0 0 1 0 0 0 1]
constraints = [M*v == b, v >= z]
objective = z
problem = Convex.maximize(objective, constraints)
solve!(problem, GLPK.Optimizer(msg_lev=GLPK.MSG_ALL ))
@printf("\n")
@printf("Optimal objective z = %f\n", problem.optval)
@printf("Optimal point     v = %s\n", v.value)
@printf("Dual variable 1   %s\n", problem.constraints[1].dual)
@printf("Dual variable 2   %s\n", problem.constraints[2].dual)


GLPK Simplex Optimizer, v4.64
21 rows, 10 columns, 42 non-zeros
      0: obj =  -0.000000000e+00 inf =   3.000e+00 (8)
      7: obj =  -2.500000000e-01 inf =   0.000e+00 (0)
*    10: obj =  -1.250000000e-01 inf =   0.000e+00 (0)
OPTIMAL LP SOLUTION FOUND

Optimal objective z = -0.125000
Optimal point     v = [0.375; 0.125; 0.125; -0.125; -0.125; 0.125; 0.125; 0.375]
Dual variable 1   [0.5; 0.0; 0.0; 0.0; -0.0; -0.5; -0.0; -0.0; -0.5; -0.0; -0.0; 0.0]
Dual variable 2   [-0.0; -0.0; -0.0; 0.5; 0.5; -0.0; -0.0; -0.0]


The optimal objective is z=-0.125 which is less than zero so the distribution is not compatible.

## Homework 2

In [46]:
#HW 2


# variable to optimize
v = Variable(8)
# slack variable
z = Variable(1)
# Contraints (M taken from the paper and b and c calculated by hand)
t = Variable(1)
b = [1/2-1/6, -2/6, -2/6, 1/2-1/6, 1/2-1/6, -2/6, -2/6, 1/2-1/6, 0, 0, 0, 0] #coefficient in front of t
c = [1/6, 2/6, 2/6, 1/6, 1/6, 2/6, 2/6, 1/6, 1/4, 1/4, 1/4, 1/4] # independant term
M = [1 1 0 0 0 0 0 0;
     0 0 1 1 0 0 0 0;
     0 0 0 0 1 1 0 0;
     0 0 0 0 0 0 1 1;
     1 0 1 0 0 0 0 0;
     0 1 0 1 0 0 0 0;
     0 0 0 0 1 0 1 0;
     0 0 0 0 0 1 0 1;
     1 0 0 0 1 0 0 0;
     0 1 0 0 0 1 0 0;
     0 0 1 0 0 0 1 0;
     0 0 0 1 0 0 0 1]

constraints = [M*v == b*t+c, v >= z, t <= 1, t >= 0] # you can change the interval for t and see if the problem is feasible or not
objective = z
problem = Convex.maximize(objective, constraints)
solve!(problem, GLPK.Optimizer(msg_lev=GLPK.MSG_ALL ))
@printf("\n")
@printf("Optimal objective z = %f\n", problem.optval)
@printf("Optimal point     v = %s\n", v.value)
@printf("Optimal parameter     t = %s\n", t.value)
@printf("Dual variable 1   %s\n", problem.constraints[1].dual)
@printf("Dual variable 2   %s\n", problem.constraints[2].dual)

GLPK Simplex Optimizer, v4.64
23 rows, 11 columns, 52 non-zeros
      0: obj =  -0.000000000e+00 inf =   3.100e+00 (13)
      9: obj =  -5.000000000e-02 inf =   0.000e+00 (0)
*    12: obj =   8.500000000e-02 inf =   4.163e-17 (0)
OPTIMAL LP SOLUTION FOUND

Optimal objective z = 0.085000
Optimal point     v = [0.08499999999999999; 0.125; 0.125; 0.16499999999999998; 0.165; 0.12499999999999999; 0.12499999999999999; 0.085]
Optimal parameter     t = 0.13
Dual variable 1   [0.0; 0.0; 0.5; 0.0; -0.0; -0.0; -0.0; -0.5; -0.5; -0.0; -0.0; 0.0]
Dual variable 2   [0.5; -0.0; -0.0; -0.0; -0.0; -0.0; -0.0; 0.5]


The problem is not feasible for 1>=t>=0.62. For 0.62>=t>=0, the problem seems feasible.

## Homework 3

In [30]:
# HW3
# variable to optimize
v = Variable(8)
# slack variable
z = Variable(1)
# Contraints (M taken from the paper and b calculated by hand)
b = [1/3, 1/3, 1/3, 0, 1/3, 1/3, 1/3, 0, (2/3)^2, 2/3*1/3, 2/3*1/3, (1/3)^2]
M = [1 1 0 0 0 0 0 0;
     0 0 1 1 0 0 0 0;
     0 0 0 0 1 1 0 0;
     0 0 0 0 0 0 1 1;
     1 0 1 0 0 0 0 0;
     0 1 0 1 0 0 0 0;
     0 0 0 0 1 0 1 0;
     0 0 0 0 0 1 0 1;
     1 0 0 0 1 0 0 0;
     0 1 0 0 0 1 0 0;
     0 0 1 0 0 0 1 0;
     0 0 0 1 0 0 0 1]
constraints = [M*v == b, v >= z]
objective = z
problem = Convex.maximize(objective, constraints)
solve!(problem, GLPK.Optimizer(msg_lev=GLPK.MSG_ALL ))
@printf("\n")
@printf("Optimal objective z = %f\n", problem.optval)
@printf("Optimal point     v = %s\n", v.value)
@printf("Dual variable 1   %s\n", problem.constraints[1].dual)
@printf("Dual variable 2   %s\n", problem.constraints[2].dual)

GLPK Simplex Optimizer, v4.64
21 rows, 10 columns, 42 non-zeros
      0: obj =  -0.000000000e+00 inf =   3.000e+00 (10)
      5: obj =  -0.000000000e+00 inf =   0.000e+00 (0)
*     9: obj =  -0.000000000e+00 inf =   0.000e+00 (0)
OPTIMAL LP SOLUTION FOUND

Optimal objective z = 0.000000
Optimal point     v = [0.1111111111111111; 0.2222222222222222; 0.2222222222222222; 0.1111111111111111; 0.3333333333333333; 0.0; 0.0; 0.0]
Dual variable 1   [0.0; 0.5; -0.5; -0.0; -0.5; -0.0; -0.0; -0.0; 0.5; -0.0; -0.0; -0.5]
Dual variable 2   [-0.0; -0.0; -0.0; -0.0; -0.0; 0.5; -0.0; 0.5]


The problem is indeed feasible as z>=0. It is still not compatible with the causality structure we test as it is not compatible with the Spiral inflation (cf. paper). 

## Homework 4

In [ ]:
# function to extend sum with semi-colon. It is used by the function marginalize below
function Base.:+(b::Colon,a::Int64)
    b
end

In [145]:
"""
Marginalize
    marginalize([0,0,:]) will give the indices to sum to obtain P{00_}
    marginalize([:,:,1,0]) will give the indices to sum to obtain P{__10}
"""
function marginalize(variables)
    n = length(variables)
    for i=1:n
        variables[i] += 1
    end
    reverse!(variables)
    size = 2*ones(Int,n)
    cube = zeros(size...)
    cube[variables...].=1
    return cube[:]
end

marginalize (generic function with 1 method)

In [183]:
# HW4

# variable to optimize
v = Variable(2^6)
# slack variable
z = Variable(1)
# constraints
M = zeros(2^4+6*4+3*2+3*8,2^6)
b = zeros(2^4+6*4+3*2+3*8)

# P_A1B1C1 = P_ABC
l=1;
for i=0:1
    for j=0:1
        for k=0:1
            M[l,:] = marginalize([i,j,k,:,:,:])
            if i+j+k==1
                b[l]=1/3
            end
            l+=1
        end
    end
end

# P_A2B2C2 = P_A P_B P_C
for i=0:1
    for j=0:1
        for k=0:1
            M[l,:] = marginalize([:,:,:,i,j,k])
            if i==0 Pi = 2/3  else  Pi = 1/3 end
            if j==0 Pj = 2/3  else  Pj = 1/3 end
            if k==0 Pk = 2/3  else  Pk = 1/3 end
            b[l] = Pi*Pj*Pk
            l+=1
        end
    end
end

# P_A2C1 == P_AC
for i=0:1
    for j=0:1
        M[l,:] = marginalize([:,:,i,j,:,:])
        if i+j != 2
            b[l] = 1/3
        end
        l+=1
    end
end

# P_A1C2 == P_A P_C
for i=0:1
    for j=0:1
        M[l,:] = marginalize([i,:,:,:,:,j])
        if i==0 Pi = 2/3  else  Pi = 1/3 end
        if j==0 Pj = 2/3  else  Pj = 1/3 end
        b[l] = Pi*Pj
        l+=1
    end
end

# P_B2A1 == P_BA
for i=0:1
    for j=0:1
        M[l,:] = marginalize([i,:,:,:,j,:])
        if i+j != 2
            b[l] = 1/3
        end
        l+=1
    end
end

# P_B1A2 == P_B P_A
for i=0:1
    for j=0:1
        M[l,:] = marginalize([:,i,:,j,:,:])
        if i==0 Pi = 2/3  else  Pi = 1/3 end
        if j==0 Pj = 2/3  else  Pj = 1/3 end
        b[l] = Pi*Pj
        l+=1
    end
end

# P_C2B1 == P_CB
for i=0:1
    for j=0:1
        M[l,:] = marginalize([:,i,:,:,:,j])
        if i+j != 2
            b[l] = 1/3
        end
        l+=1
    end
end

# P_C1B2 == P_C P_B
for i=0:1
    for j=0:1
        M[l,:] = marginalize([:,:,i,:,j,:])
        if i==0 Pi = 2/3  else  Pi = 1/3 end
        if j==0 Pj = 2/3  else  Pj = 1/3 end
        b[l] = Pi*Pj
        l+=1
    end
end

# P_A2 == P_A
for i=0:1
        M[l,:] = marginalize([:,:,:,i,:,:])
        if i==0 Pi = 2/3  else  Pi = 1/3 end
        b[l] = Pi
        l+=1
end

# P_B2 == P_B
for i=0:1
        M[l,:] = marginalize([:,:,:,:,i,:])
        if i==0 Pi = 2/3  else  Pi = 1/3 end
        b[l] = Pi
        l+=1
end

# P_C2 == P_C
for i=0:1
        M[l,:] = marginalize([:,:,:,:,:,i])
        if i==0 Pi = 2/3  else  Pi = 1/3 end
        b[l] = Pi
        l+=1
end

# P_A2C1B2 == P_AC P_B
for i=0:1
    for j=0:1
        for k=0:1
            M[l,:] = marginalize([:,:,i,j,k,:])
            if i + j!=2 Pij = 1/3  else  Pij = 0 end
            if k==0 Pk = 2/3  else  Pk = 1/3 end
            b[l] = Pij*Pk
            l+=1
        end
    end
end

# P_A1B1C2 == P_AB P_C
for i=0:1
    for j=0:1
        for k=0:1
            M[l,:] = marginalize([i,j,:,:,:,k])
            if i + j!=2 Pij = 1/3  else  Pij = 0 end
            if k==0 Pk = 2/3  else  Pk = 1/3 end
            b[l] = Pij*Pk
            l+=1
        end
    end
end

# P_B1C1A2 == P_BC P_A
for i=0:1
    for j=0:1
        for k=0:1
            M[l,:] = marginalize([:,i,j,k,:,:])
            if i + j!=2 Pij = 1/3  else  Pij = 0 end
            if k==0 Pk = 2/3  else  Pk = 1/3 end
            b[l] = Pij*Pk
            l+=1
        end
    end
end

constraints = [M*v == b, v >= z]
objective = z
problem = Convex.maximize(objective, constraints)
solve!(problem, GLPK.Optimizer(msg_lev=GLPK.MSG_ALL ))
@printf("\n")
@printf("Optimal objective z = %f\n", problem.optval)
@printf("Optimal point     v = %s\n", v.value)
@printf("Dual variable 1   %s\n", problem.constraints[1].dual)
@printf("Dual variable 2   %s\n", problem.constraints[2].dual)

GLPK Simplex Optimizer, v4.64
135 rows, 66 columns, 1026 non-zeros
      0: obj =  -0.000000000e+00 inf =   1.400e+01 (56)
     22: obj =  -0.000000000e+00 inf =   4.667e+00 (42)
LP HAS NO PRIMAL FEASIBLE SOLUTION

Optimal objective z = 0.000000
Optimal point     v = nothing
Dual variable 1   [NaN; 1.0; 1.0; NaN; 1.0; NaN; NaN; NaN; NaN; 1.0; 1.0; 1.0; NaN; 1.0; 1.0; 1.0; 1.0; 1.0; 1.0; NaN; NaN; 1.0; 1.0; NaN; NaN; 1.0; 1.0; NaN; 1.0; 1.0; 1.0; 1.0; NaN; 1.0; 1.0; NaN; 1.0; 1.0; NaN; 1.0; 1.0; 1.0; 1.0; 1.0; 1.0; 1.0; 1.0; NaN; 1.0; 1.0; NaN; 1.0; NaN; NaN; NaN; 1.0; NaN; 1.0; 1.0; NaN; NaN; NaN; 1.0; NaN; NaN; 1.0; 1.0; 1.0; NaN; NaN]
Dual variable 2   [NaN; NaN; NaN; NaN; NaN; NaN; NaN; NaN; NaN; NaN; NaN; NaN; NaN; NaN; NaN; NaN; NaN; NaN; NaN; NaN; NaN; NaN; NaN; NaN; NaN; NaN; NaN; NaN; NaN; NaN; NaN; NaN; NaN; NaN; NaN; NaN; NaN; NaN; NaN; NaN; NaN; NaN; NaN; NaN; NaN; NaN; NaN; NaN; NaN; NaN; NaN; NaN; NaN; NaN; NaN; NaN; NaN; NaN; NaN; NaN; NaN; NaN; NaN; NaN]


┌ Warning: Problem status INFEASIBLE; solution may be inaccurate.
└ @ Convex /home/bdebruyne/.julia/packages/Convex/kMsJj/src/solution.jl:229


We see that the problem is infeasible for the spiral inflation.